### PR TITLE
feat: getContract

### PR DESCRIPTION
**Pull request by tmm**
Fully type-safe contract instances

#165 

Methods

image

Autocomplete function/event names

CleanShot 2023-03-24 at 13 14 15@2x

Type-safe arguments (as usual 😊)

image

No public client, no access to public actions

image

No wallet client, no access to wallet actions

image

No account attached to wallet client, required for wallet actions

image

### CODE REVIEW

File: src/actions/getContract.ts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+export type GetContractReturnType<
+  TAbi extends Abi | readonly unknown[] = Abi,
+  TPublicClient extends PublicClient | unknown = unknown,
+  TWalletClient extends WalletClient | unknown = unknown,
+  _EventNames extends string = TAbi extends Abi
+    ? Abi extends TAbi
+      ? string
+      : ExtractAbiEventNames<TAbi>
+    : string,
```
Comment (roninjin10) [inline src/actions/getContract.ts:58]: Am I reading this right?   You are basically saying if TAbi is unknown it's string.   If TABI is literally the wide type ABI it's string too (this is a corner case).   And then in happy path extract event names.

File: src/types/abi.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ * @example
+ * import { seaportAbi } from 'abitype/test'
+ * type Result = IsInferrableAbi<typeof seaportAbi>
+ * //   ^? true
+ */
+export type IsInferrableAbi<TAbi extends Abi | readonly unknown[]> = IsNever<
+  IsAbi<TAbi> & (Abi extends TAbi ? false : true)
+> extends true
+  ? false
+  : true
```
[the remark this answers is not in the document]
Comment (tmm) [reply src/types/abi.ts:16]: Thinking about it some more: Probably best for this to be a type utility called `IsInferrable<TType, TValue>`

**Comment by tmm** [comment]
benched

```bash
❯ pnpm bench src/actions/getContract.bench.ts

 DEV  v0.29.2 /Users/tom/Code/viem

 ✓ src/actions/getContract.bench.ts (3) 1791ms
   ✓ Contract instance (3) 1789ms
     · viem: `getContract`       1,296,595.34 ops/sec ±2.20% (648298 samples) fastest
     · ethers@5: `new Contract`      9,624.43 ops/sec ±0.89% (  4813 samples)
     · ethers@6: `new Contract`      5,745.53 ops/sec ±0.62% (  2873 samples) slowest


 BENCH  Summary

  viem: `getContract` - src/actions/getContract.bench.ts > Contract instance
    134.72x faster than ethers@5: `new Contract`
    225.67x faster than ethers@6: `new Contract`
```

**Comment by roninjin10** [comment]
Using viem for your contracts 
image

File: src/actions/getContract.ts [hunk 3]
```diff
@@ -0,0 +1,706 @@
...
+ * @internal exporting for testing only
+ */
+export function getFunctionArgsAndParams(
+  rest: [args?: readonly unknown[], params?: object],
+) {
+  const hasArgs = rest.length && Array.isArray(rest[0])
+  const args = hasArgs ? rest[0]! : []
+  const params = (hasArgs ? rest[1] : rest[0]) ?? {}
+  return { args, params }
+}
```
Comment (jxom) [inline src/actions/getContract.ts:445]: Can we add some test cases to this fn?

File: site/docs/contract/getContract.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        type: 'uint256',
+      },
+    ],
+  },
+  ...
+] as const;
+```
+:::
+
+Using Contract Instances can make it easier to work with contracts if you don't want to pass the `abi` and `address` properites every time you perform contract actions, e.g. [`readContract`](/docs/contract/readContract.html), [`writeContract`](/docs/contract/writeContract.html), [`estimateContractGas`](/docs/contract/estimateContractGas.html), etc. For example, you can convert the following individual calls:
```
Comment (jxom) [inline site/docs/contract/getContract.md:89]: Maybe a callout at the end of this paragraph to suggest to the consumer to switch between the tabs to see the difference?

```suggestion
Using Contract Instances can make it easier to work with contracts if you don't want to pass the `abi` and `address` properites every time you perform contract actions, e.g. [`readContract`](/docs/contract/readContract.html), [`writeContract`](/docs/contract/writeContract.html), [`estimateContractGas`](/docs/contract/estimateContractGas.html), etc. Switch between the tabs below to see the difference between standalone Contract Actions and Contract Instance Actions:
```

File: site/docs/contract/getContract.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+import { wagmiAbi } from './abi'
+import { publicClient } from './client'
+
+// 1. Create contract instance
+const contract = getContract({
+  address: '0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2',
+  abi: wagmiAbi,
+  publicClient,
+})
+// 2. Call contract methods, listen to events, etc.
```
Comment (jxom) [inline site/docs/contract/getContract.md:41]:
```suggestion
})

// 2. Call contract methods, listen to events, etc.
```

File: site/docs/contract/getContract.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+})
+const filter = await publicClient.createContractEventFilter({
+  address: '0xfba3912ca04dd458c843e2ee08967fc04f3579c2',
+  abi: wagmiAbi,
+  eventName: 'Transfer',
+  args: {  
+    from: '0xd8da6bf26964af9d7eed9e03e53415d37aa96045',
+    to: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac'
+  }
+})
```
Comment (jxom) [inline site/docs/contract/getContract.md:116]: Should we replace this with `watchContractEvent`? Feel like contract event listeners might be used more than filters.

File: site/docs/contract/getContract.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+### abi
+
+- **Type:** [`Abi`](/docs/glossary/types#abi)
+
+The contract's ABI.
+
+### publicClient
+
+- **Type:**
```
Comment (jxom) [inline site/docs/contract/getContract.md:177]: Does there need to be a type here?

File: site/docs/contract/getContract.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+- [`write`](/docs/contract/writeContract.html)
+
+## Parameters
+
+### address
+
+- **Type:** [`Address`](/docs/glossary/types#address)
+
+The contract address.
+
```
Comment (jxom) [inline site/docs/contract/getContract.md:168]: Should we include example snippet here?

**Review by jxom — APPROVED** [review]
Really solid stuff!

**Review by jxom — APPROVED** [review]
